src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupLinearFunctional.java

```diff
@@ -0,0 +1,516 @@
+
+			// We simply compute a matrix multiplication chain in coefficient space, i.e.,
+			// 					t(L) %*% R = t(coeff(L)) %*% W %*% coeff(R)
+			// where W is a weight matrix capturing the size of the shared dimension (weightMatrix above)
+			// and coeff(X) denotes the 2 x n matrix of the m x n matrix X.
+			MatrixBlock tmp = new MatrixBlock(lhs.getNumCols(), 2, false);
+			LibMatrixMult.matrixMult(coefficientMatrixLhs, weightMatrix, tmp);
+			LibMatrixMult.matrixMult(tmp, coefficientMatrixRhs, tmpRet);
+		} else if(lhs instanceof APreAgg) {
+			// TODO: implement
+			throw new NotImplementedException();
+		} else {
+			throw new NotImplementedException();
+		}
+
+		final double[] resV = result.getDenseBlockValues();
+		if(tmpRet.isEmpty())
+			return;
+		else if(tmpRet.isInSparseFormat()) {
+			SparseBlock sb = tmpRet.getSparseBlock();
+			for(int row = 0; row < lhs._colIndexes.length; row++) {
+				if(sb.isEmpty(row))
+					continue;
+				final int apos = sb.pos(row);
+				final int alen = sb.size(row) + apos;
+				final int[] aix = sb.indexes(row);
+				final double[] avals = sb.values(row);
+				final int offRes = lhs._colIndexes[row] * result.getNumColumns();
+				for(int col = apos; col < alen; col++)
+					resV[offRes + _colIndexes[aix[col]]] += avals[col];
+			}
+		}
+		else {
+			double[] tmpRetV = tmpRet.getDenseBlockValues();
+			for(int row = 0; row < lhs.getNumCols(); row++) {
+				final int offRes = lhs._colIndexes[row] * result.getNumColumns();
+				final int offTmp = row * getNumCols();
+				for(int col = 0; col < getNumCols(); col++) {
+					resV[offRes + _colIndexes[col]] += tmpRetV[offTmp + col];
+				}
+			}
+		}
+	}
+
+	@Override
+	public void tsmmAColGroup(AColGroup other, MatrixBlock result) {
+		throw new DMLCompressionException("Should not be called");
+	}
+
+	@Override
+	protected AColGroup sliceSingleColumn(int idx) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	protected AColGroup sliceMultiColumns(int idStart, int idEnd, int[] outputCols) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public AColGroup copy() {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public boolean containsValue(double pattern) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public long getNumberNonZeros(int nRows) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public AColGroup replace(double pattern, double replace) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public void readFields(DataInput in) throws IOException {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public void write(DataOutput out) throws IOException {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public long getExactSizeOnDisk() {
+		long ret = super.getExactSizeOnDisk();
+		ret += _coefficents[0].length * _coefficents.length * 8L;
+		ret += 4L; // _numRows
+		return ret;
+	}
+
+	@Override
+	protected void computeProduct(double[] c, int nRows) {
+		c[0] = 1;
```

Comment:
do not set it to 1, since you have to consider the input c as already allocated correctly.
Instead you can add an early abort `if 0 return`

Reply:
furthermore you could add a speedup, by adding a containsZero method that detects if any of the values contained in the function is zero return zero